toolkit/tools/graphpkgfetcher/graphpkgfetcher.go

```diff
@@ -124,20 +132,94 @@ func fetchPackages() (err error) {
+	if *tryDownloadDeltaRPMs {
+		logger.Log.Info("Attempting to download delta RPMs for build nodes")
+		err = downloadDeltaNodes(dependencyGraph, cloner)
+		if err != nil {
+			err = fmt.Errorf("failed to download delta RPMs:\n%w", err)
+			return
+		}
+	}
+
+	// Write the final graph to file
 	err = pkggraph.WriteDOTGraphFile(dependencyGraph, *outputGraph)
 	if err != nil {
 		err = fmt.Errorf("failed to write cache graph to file:\n%w", err)
 		return
 	}
 
 	// If we grabbed any RPMs, we need to convert them into a local repo
-	if hasUnresolvedNodes {
+	if *tryDownloadDeltaRPMs || hasUnresolvedNodes {
 		err = cloner.ConvertDownloadedPackagesIntoRepo()
 		if err != nil {
 			err = fmt.Errorf("failed to convert downloaded RPMs into a repo:\n%w", err)
 			return err
 		}
+
+		if strings.TrimSpace(*outputSummaryFile) != "" {
+			err = repoutils.SaveClonedRepoContents(cloner, *outputSummaryFile)
+			if err != nil {
+				err = fmt.Errorf("failed to save cloned repo contents:\n%w", err)
+				return err
+			}
+		}
+	}
+	return
+}
+
+// downloadDeltaNodes will look at the final cached graph we saved and see if any RPMS can be download instead of built.
+// If the previous part of the fetcher worked well we should be able to download only the delta RPMs we need
+// to build our packages or image (i.e. we should be able to create a subgraph just like we would for the build step)
+//   - dependencyGraph: The graph to use to find the packages we need to build. Should have any caching operations already
+//     performed on it. Will be updated with the paths to the delta RPMs we download.
+//   - cloner: The cloner to use to download the RPMs
+//
+// We also access the package list related globals:
+//   - pkgsToBuild: The list of packages to build
+//   - pkgsToRebuild: The list of packages to rebuild
+//   - pkgsToIgnore: The list of packages to ignore
+//   - imageConfig: The image config to use to find the packages we need to build
+//   - baseDirPath: The base directory to use to find the packages we need to build
+func downloadDeltaNodes(dependencyGraph *pkggraph.PkgGraph, cloner *rpmrepocloner.RpmRepoCloner) (err error) {
+	const (
+		useImplicitForOptimization = true
+	)
+
+	timestamp.StartEvent("delta package download", nil)
+	defer timestamp.StopEvent(nil)
+
+	// Generate the list of packages that need to be built. If none are requested then all packages will be built. We
+	// don't care about explicit rebuilds here since we are going to rebuild them anyway.
+	packageVersToBuild, _, _, err := schedulerutils.ParseAndGeneratePackageList(dependencyGraph, *pkgsToBuild, *pkgsToRebuild, *pkgsToIgnore, *imageConfig, *baseDirPath)
+	if err != nil {
+		err = fmt.Errorf("unable to generate package build list to calculate delta downloads: %w", err)
+		return
+	}
+
+	// We will heavily modify this graph so it should not be used for anything else, create a copy of it to work with.
+	deltaPkgGraphCopy, err := dependencyGraph.DeepCopy()
```

Comment:
>rr := dependencyGraph.DeepCopy() [](http://example.com/codeflow?start=21&length=32)

I'll need you to guide me through the changes in this file - it's too late at night now.;)